### ISSUE TITLE
Able to build capabilities after/while discovery process.

### ISSUE THREAD

**Issue by nklerk** [NEEOInc/neeo-sdk#143]
While the idea was allready requested (https://github.com/NEEOInc/neeo-sdk/issues/53) and denied i want to re request it in a more specific way. as i believe there where some misunderstandings.

With some of the drivers I have written i'm running into the issue that i want to process a descovery and within that discovery get and build the capabilities of a device. I'm just about to start my next project where again this is important, and here is why:

Some products have different models. like for instance lifx lights. now we could define every specific light and its capabilities but this needs the driver to include the device types all statically defined while this product is very capable of returning all the capabilities it has. When a new light is being introduced it is automatically supported with the request I'm trying to make here.

The reason i'm asking it again is because of the folowing project i'm about to start. I'm starting to work on a HDAnywhere driver, There are a static set of devices until a new device is introduced. but besides the available devices these devices can also be stacked and thus need to offer a different capability set.

Offcorse I could discover devices at the start of the driver but this would only be a crappy workaround. I don't want to only have discovery availeble at the start of the script for the folowing reasons.
- The NEEO UI offers discovery and doesn't make sense to do discovery at the start of the driver. (can't explain that to a user)
- When drivers are added to the brain a user needs te reboot the brain to discover a device.
- If the discovery failed for what ever reason the full driver has to be restarted.
- If multiple devices are at a user's site a restart of the driver would mean all devices need to be available to otherwise one driver fails.

Process:

- Build a basic device set. julst as it's now but with common capabilities (think of Power on Power off)
- The driver discovers all devices and checks what the device is capable of. (example. how many inputs does the device have and add buttons for the exact amount of inputs)
- The discovery process can return devices but also device specific capabilities.
- When the user select's its driver the neeo gets the common and specific capabilities offered.

- In between discoveries (in case the driver has been restarted) capabilities can be added.

**Comment by pfiaux** [comment]
We're working on a dynamic device feature to be included in the next release, I think it will cover the majority of your needs. Here's a quick overview:

* You can define a device as dynamically built (you only define an "abstract" device with no capabilities/macros/...). Dynamic devices **must** use discovery.
* During discovery for every result you have you can build them (macros, capabilities, ...) like you would for a non dynamic device on startup
* These are returned to the Brain and the added result will be based on the specific result selected.

So basically your discovery results can cover different devices.

Regarding your last point, changing a device after it's been added on the Brain, I think this will be covered through another upcoming feature which is driver versioning #18

I think the combination of these 2 features will cover most of the process you mentioned above. Do you see any needs that wouldn't be covered through this?

**Comment by nklerk** [comment]
This is exactly what I would like to see. And looks like it fully covers what I’m requesting.

**Comment by neophob** [comment]
Thanks for the feedback @nklerk. I'll close it then
